Inject backstopTools

laubsauger opened the pull request:
Solves #541.

I would be glad if somebody could enlighten me why this works. 

Using chromy's `inject` method resulted in no joy at all. Don't get why. Worked in other cases for me.
Trying to do this via `chromy.on('Page.loadEventFired')...` also didn't work.
I guess because the first evaluations are using these methods before this event?

garris commented on core/util/runChromy.js:
```diff
@@ -9,114 +9,16 @@ const CHROMY_STARTING_PORT_NUMBER = 9222;
+const injectBackstopTools = require('../../capture/backstopTools.js');
 const BackstopException = require('../util/BackstopException.js');
 
-// ============================== MOVE TO NEW FILE VVVVV ===========================
```
Please move these functions into the new file include verbatim -- these have been tested.  Let's not use the original backstopTools methods.  Thanks.

laubsauger replied:
I'm not sure i understand your request. You want those methods from `runChromy.js` to be moved to `backstopTools.js` as is? Isn't that the case? I saw no difference aside from the usage of constants in for the original methods in `backstopTools.js`.

garris replied:
Thanks for looking at this.  Yes, That is the request.   I think there may have been tweaks to the methods inside of `runChromy.js`  -- That code has been working in prod -- so I'd like to ensure there are as few changes as possible.    I think there may be some minimal inline documentation added as well.   Simply copying the whole thing may be easier than attempting to diff.

garris replied:
We may also want to extend the names to avoid any collisions.  I never liked the name .exists -- it's way too common.  maybe `window._backstopTools.exists()` would be better.

garris commented on core/util/runChromy.js:
```diff
@@ -347,6 +232,8 @@ function processScenarioView (scenario, variantOrScenarioLabelSafe, scenarioLabe
   }
 
   return new Promise((resolve, reject) => {
+    injectBackstopTools(chromy);
```
injectBackstopTools() should probably be moved to line 211 -- and only run if `!window.exists`.   Does that work?

garris replied:
injectBackstopTools() probably also unconditionally needs to run on line 151.

laubsauger commented:
You're right. This kills readyEvent/readySelector. Slight oversight on my part.

garris commented:
No worries!

laubsauger commented:
I've moved the injection to the suggested places, moved backstopTools method as is to a new property to avoid conflicts.
Your test config is working now.

laubsauger commented on core/util/runChromy.js:
```diff
@@ -247,30 +149,15 @@ function processScenarioView (scenario, variantOrScenarioLabelSafe, scenarioLabe
-  // chromy.evaluate(() => console.log("window.exists ^^^^"))
-
-// =================UPDATE ^^^^^^^^^ ==================
+  injectBackstopTools(chromy);
```
I've tried to do it conditionally but it failed then.

````javascript
chromy.evaluate(() => {
    return !window._backstopTools;
}).result(_result => {
    if (_result) {
      injectBackstopTools(chromy)
    }
});
````
I don't have time left today to figure this out.

garris commented:
This is awesome.  Thank you!
